Add Tokenized Vaults Standard TEP

ipromise2324 opened the pull request:
This TEP establishes a standardized interface for tokenized vaults on TON, enabling consistent deposit/withdrawal operations, rate queries, and cross-protocol communication. Inspired by ERC-4626 but adapted to TON's asynchronous design, it improves DeFi ecosystem interoperability and reduces integration complexity.

ipromise2324 committed:
Add vault standard TEP

ipromise2324 committed:
Update TEP identifier

ipromise2324 committed:
refactor: rename vault standard files 0000 → 0524

andrew-cong commented:
This proposal represents a strong, forward-thinking contribution to TON's DeFi ecosystem, addressing key fragmentation issues while thoughtfully adapting established standards like ERC-4626 to our asynchronous, message-driven architecture. It has high potential to accelerate protocol composability and developer adoption.

ipromise2324 committed:
Update asset image paths in vault standard documentation to reflect new directory structure

SlorrUndef commented:
Hey @ipromise2324, thanks a lot for sharing - your proposal looks really solid! 

We’ve also been working on an implementation that already includes feedback from top-tier partners like Aave, Ethena, and Affluent. Given that it reflects inputs from several ecosystem players, it feels more practical to move forward with this version as a base. That said, your proposal brings in great ideas, and it would be super valuable if you could comment on the current draft so we can capture the best of both approaches. 

https://github.com/ton-blockchain/TEPs/pull/525

We’ll also make sure to provide more transparency on future standards in progress, so everyone can stay aligned.

Looking forward to your feedback!

alan890104 commented:
@ipromise2324 well done, nice comparison

ipromise2324 committed:
Add TL-B specifications for vault operations including deposits, withdrawals, and event logging in vault standard documentation

ipromise2324 committed:
Add detailed guidelines for implementing the vault standard, including use cases

shuva10v commented on text/0524-vault-standard.md:
```diff
@@ -0,0 +1,949 @@
+
+![deposit-ton](../assets/0524-vault-standard/deposit-ton.png)
+
+- **Description**: Mint shares to `receiver` by depositing exactly `depositAmount` of TON.
```
it would be great to specify TL-B for minting shares. TEP-74 doesn't specify it but most of the implementations are using the same approach - internal_transfer message sent from master to jetton wallet address:

```
internal_transfer  query_id:uint64 amount:(VarUInteger 16) from:MsgAddress
                     response_address:MsgAddress
                     forward_ton_amount:(VarUInteger 16)
                     forward_payload:(Either Cell ^Cell)
                     = InternalMsgBody;
```

Having the same format for all mints streamlines indexers/explorers integration

shuva10v commented on text/0524-vault-standard.md:
```diff
@@ -0,0 +1,949 @@
+  | `receiver` | `Address` | Address receiving shares. |
+  | `depositAsset` | Cell<[Asset](#asset)>? | Deposited asset (required for multi-asset vaults, null for single asset). |
+  | `depositAmount` | `Coins` | Deposited asset amount. |
+  | `shares` | `Coins` | Minted shares. |
```
Since underlying amount is subject to change it would be great to add `totalAssetAmount` in the log event. It will allow to reconstruct the vault TVL simply by accessing all the logs messages without dealing with get methods invocations.

ipromise2324 committed:
Enhance vault standard documentation with improved formatting, additional guidelines for share minting, and detailed specifications for deposit and withdrawal events.

ipromise2324 committed:
Add VaultStateAfter structure to vault standard documentation, detailing its fields and usage in deposit, withdrawal, and quote operations.

blendotsequad reviewed (commented):
God

ipromise2324 committed:
Update ExtraCurrencyId and tokenId types in vault standard documentation from uint32 to int32 for consistency and accuracy.

ipromise2324 committed:
Add definition of Vault functionality to the vault standard documentation

ipromise2324 committed:
Refactor VaultStateAfter to VaultState and update field naming for consistency

- Rename VaultStateAfter to VaultState throughout the specification
- Rename totalAssetAmount to totalAssets for clarity
- Update Quoted event to include totalSupply and normalizedTotalAssets instead of vaultStateAfter for better event structure
- Fix table formatting alignment in Asset type definition
